staging/src/k8s.io/kubectl/pkg/drain/filters.go

```diff
@@ -187,13 +187,15 @@ func (d *Helper) daemonSetFilter(pod corev1.Pod) PodDeleteStatus {
 		return MakePodDeleteStatusOkay()
 	}
 
-	if _, err := d.Client.AppsV1().DaemonSets(pod.Namespace).Get(context.TODO(), controllerRef.Name, metav1.GetOptions{}); err != nil {
-		// remove orphaned pods with a warning if --force is used
-		if apierrors.IsNotFound(err) && d.Force {
-			return MakePodDeleteStatusWithWarning(true, err.Error())
+	// No need to check if the pod is orphaned when `--force=false`
+	if d.Force {
+		if _, err := d.Client.AppsV1().DaemonSets(pod.Namespace).Get(context.TODO(), controllerRef.Name, metav1.GetOptions{}); err != nil {
+			// remove orphaned pods with a warning if --force is used
+			if apierrors.IsNotFound(err) {
+				return MakePodDeleteStatusWithWarning(true, err.Error())
+			}
+			return MakePodDeleteStatusWithError(err.Error())
```

Comment:
There are two subtle additional behaviour changes in this PR caused by the relocation of this line under `d.Force` and resultant inability to distinguish between standalone and orphaned pods without forcing enabled: 
* If force is false _and_ the pod is orphaned, previously it would fail with an error message about the parent not being found, and now it will fail with an error message about needing to ignore daemonsets (from L202). 
* If force is false _and_ daemonsets are ignored _and_ the pod is orphaned, previously it would fail with an error message about the parent not being found, and now it will succeed with a message about the pod being ignored because it is DS-managed.

Wdyt about those changes? The second is somewhat concerning, because it exits successfully despite leaving a pod running that is not actually managed by the DaemonSet controller--normally such pods cause the command to fail (unless `--force` is used of course). I'm cautiously leaning towards this being acceptable, however.

It is possible to cover these with the existing unit tests, which unfortunately are not validly covering the orphan case at all anymore. It seems the test case originally predated ownerReferences and was not updated correctly when they were introduced. Here is that correction + test cases that pass on master and need adjustments to pass with this PR: https://github.com/kubernetes/kubernetes/pull/41864/files#diff-02a1d45d63c5a98f29bf091157cd2080fe8b7134bae5c45cd5dc42e1e2a2eda2. PTAL.

Reply:
Try to summarize: 

force | ignore-ds | orphaned | can-get-DS | current result |  correctness | should be
-- | -- | -- | --  | --  | -- | --  
false | false | false | true | MakePodDeleteStatusWithError(daemonSetFatal) |  ✅ | 
false | false | true | true | MakePodDeleteStatusWithError(err.Error()) | ✅ | 
false | true | false | true | **warning** MakePodDeleteStatusWithWarning(false, daemonSetWarning) | ✅ | 
false | true | true | true | MakePodDeleteStatusWithError(err.Error()) | ✅ | 
true | false | false | true | MakePodDeleteStatusWithError(daemonSetFatal) | ✅ | 
true | false | true | true | **Delete** MakePodDeleteStatusWithWarning(true, err.Error()) | ✅ | 
true | true | false | true | **warning** MakePodDeleteStatusWithWarning(false, daemonSetWarning) | ✅ | 
true | true | true | true | **Delete** MakePodDeleteStatusWithWarning(true, err.Error()) | ✅ | 
false | false | false | false | MakePodDeleteStatusWithError(err.Error()) |✅ | 
false | false | true | false | MakePodDeleteStatusWithError(err.Error()) | ✅| 
false | true | false | false | MakePodDeleteStatusWithError(err.Error()) |✅ | 
false | true | true | false | MakePodDeleteStatusWithError(err.Error()) |✅ | 
true | false | false | false | MakePodDeleteStatusWithError(err.Error()) |✅ | 
true | false | true | false |MakePodDeleteStatusWithError(err.Error()) | ✅| 
true | true | false | false | MakePodDeleteStatusWithError(err.Error())| ❌ |  **warning** MakePodDeleteStatusWithWarning(false, daemonSetWarning) 
true | true | true | false | MakePodDeleteStatusWithError(err.Error()) | ❌  | **warning** MakePodDeleteStatusWithWarning(false, daemonSetWarning) 

We can say that `only if force is true and the pod is orphaned,  the pod can be removed.` For all other cases, the warning/error messages may be different, but the pod will not be removed.

So if `kubectl` has no permission to get DS, it cannot tell if this is an orphaned pod. 
Only if --force and --ignore-daemonsets are both true, we should return a warning only. 

I updated the PR.

Reply:
I'm confused by the latest update; it no longer changes the permissions required to run the command, per the linked issue. What do you think of the two changes I asked about in my previous comment specifically?

Please also fix/add to the existing unit test suite as shown in the commit I shared: https://github.com/kubernetes/kubernetes/pull/41864/files#diff-02a1d45d63c5a98f29bf091157cd2080fe8b7134bae5c45cd5dc42e1e2a2eda2. I'm fine with having tests directly on the filter as well, but I prefer the coverage from `TestDrain`, because it shows the final result end users will experience, which is the result of multiple filters being applied.